projects/permissions/src/services/permissions.service.ts

```diff
@@ -1,6 +1,51 @@
-import {Injectable} from '@angular/core';
+import {Inject, Injectable} from '@angular/core';
+import {from, fromEvent, Observable} from 'rxjs';
+import {shareReplay, switchMap, tap} from 'rxjs/operators';
+import {PERMISSIONS} from '../tokens/permissions';
+import {PERMISSIONS_SUPPORT} from '../tokens/permissions-support';
+
+export type PermissionsQueryArgs = Parameters<typeof Permissions.prototype.query>[0];
 
 @Injectable({
     providedIn: 'root',
 })
-export class PermissionsService {}
+export class PermissionsService {
+    constructor(
+        @Inject(PERMISSIONS) private readonly permissions: Permissions,
+        @Inject(PERMISSIONS_SUPPORT) private permissionsSupported: boolean,
+    ) {}
+
+    state(name: PermissionName): Observable<PermissionState>;
+    state(descriptor: PermissionsQueryArgs): Observable<PermissionState>;
+    state(
+        nameOrDescriptor: PermissionName | PermissionsQueryArgs,
+    ): Observable<PermissionState> {
+        const descriptor: PermissionDescriptor =
+            typeof nameOrDescriptor === 'string'
+                ? {name: nameOrDescriptor}
+                : nameOrDescriptor;
+
+        return new Observable<PermissionState>(subscriber => {
+            if (!this.permissionsSupported) {
+                subscriber.error('Permissions is not supported in your browser');
+
+                return;
+            }
+
+            const sub = from(this.permissions.query(descriptor))
```

Comment:
How about something like that?

```ts
    return from(
        this.permissions.query(descriptor).pipe(
            switchMap(status =>
                fromEvent(status, 'change').pipe(
                    startWith(null),
                    map(() => status.state),
                ),
            ),
        ),
    ).subscribe(subscriber);
```

`PermissionState` is live so we can use `map(() => status.state)` to return actual state without typecasting event target. We can also just pass `subscriber`  to `subscribe` this way and return resulting `Subscription` as `TeardownLogic`.

Reply:
sneaky! Good thinking. I'll change to that.